### DISCUSSION TITLE
Why App::setup isn't called on startup?

### DISCUSSION THREAD

**Discussion by Kesanov** [emilk/egui#718] [Q&A]
I run the application with `eframe::run_native(Box::new(app), default());`. However, the code inside `setup` isn't called:
```rust
    #[cfg(feature = "persistence")]
    fn setup(&mut self, _ctx: &egui::CtxRef, frame: &mut Frame<'_>, storage: Option<&dyn Storage>) {
        println!("SETUP");
    }
 ```
Am I missing something?

**Comment by emilk** [comment]
Have you enabled the `persistence` feature or tried removing `#[cfg(feature = "persistence")]`?

**Reply by Kesanov** [reply]
Turns out I don't have it enabled. My `cargo.toml` contains `eframe = { version = "0.14", features = ["persistence"] }`, but it does not seem to have effect.

**Reply by manokara** [reply]
That will enable `persistence` on eframe, but not in your app. You need to declare the feature in your app as well or drop the `#[cfg]`, unless you're also planning to make persistence (saving and loading data) optional.

In that case, add this to your Cargo.toml:

```toml
[features]
persistence = []
```

And run the app with `cargo run --features persistence`, or add a `default = ["persistence"]` line to the features section.

**Reply by emilk** [reply]
`egui_template` is a bit confusing here. Fixed: https://github.com/emilk/egui_template/commit/100b9eebdd20e6b300c77afd2541afd71f0dcef9

**Reply by Kesanov** [reply]
Oh, this is the first time I see this approach. I didn't know that putting `persistence` into `default` enables it.

Thank you!

**Reply by manokara** [reply]
Also, you could remove `features = ["persistence"]` off of eframe's definition, and put it into your app's `persistence` feature, like this:

```toml
[features]
default = ["persistence"]
persistence = ["eframe/persistence"]
```

That way persistence on eframe will only be set if it is also set in your app.
